File: 프로그래머스/3/42579. 베스트앨범/베스트앨범.py
import heapq
from collections import defaultdict

def solution(genres, plays):
    answer = []
    
    genre_count = {} # 장르명: 총 재생 횟수
    genre_plays = defaultdict(list) # 장르명: [(-재생 횟수, 고유번호), ...]
    for i in range(len(plays)):
        if genres[i] in genre_count:
            genre_count[genres[i]] += plays[i]
        else:
            genre_count[genres[i]] = plays[i]
        genre_plays[genres[i]].append((-plays[i], i))
    
    genre_count_order = []
    heapq.heapify(genre_count_order)
    for k, v in genre_count.items():
        counts = (-v, k)
        heapq.heappush(genre_count_order, counts)
    
    while genre_count_order:
        _, genre = heapq.heappop(genre_count_order)
        g_plays = genre_plays[genre]
        heapq.heapify(g_plays)
        for _ in range(2):
            if g_plays:
                _, play = heapq.heappop(g_plays)
                answer.append(play)
    
    return answer

File: 프로그래머스/3/42579. 베스트앨범/베스트앨범.java
import java.util.*;

class Solution {
    static class Song {
        int idx;
        int play;

        Song(int idx, int play) {
            this.idx = idx;
            this.play = play;
        }
    }
    
    public int[] solution(String[] genres, int[] plays) {
        List<Integer> answer = new ArrayList<>();
        
        Map<String, Integer> genreCount = new HashMap<>(); // 장르: 총 재생 횟수
        Map<String, List<Song>> genrePlays = new HashMap<>(); // 장르: [Song(고유번호, 재생횟수), ...]
        
        for (int i = 0; i < plays.length; i++) {
            genreCount.put(genres[i], genreCount.getOrDefault(genres[i], 0) + plays[i]);
            genrePlays.computeIfAbsent(genres[i], k -> new ArrayList<>()).add(new Song(i, plays[i]));
        }
        
        // 속한 노래가 많이 재생된 장르 순으로 정렬
        List<String> sortedGenres = new ArrayList<>(genreCount.keySet());
        sortedGenres.sort((genre1, genre2) -> genreCount.get(genre2) - genreCount.get(genre1));
        
        for (String genre: sortedGenres) {
            List<Song> songs = genrePlays.get(genre);
            songs.sort((song1, song2) -> {
                if (song2.play != song1.play) return song2.play - song1.play; // 재생 수 내림차순
                return song1.idx - song2.idx; // 고유번호 오름차순
            });
            
            for (int i = 0; i < Math.min(2, songs.size()); i++) {
                answer.add(songs.get(i).idx);
            }
        }
        
        return answer.stream().mapToInt(i -> i).toArray();
    }
}